server/src/main/java/org/apache/calcite/server/ServerDdlExecutor.java

```diff
@@ -281,41 +281,54 @@ public void execute(SqlCreateFunction create,
    * {@code DROP VIEW} commands. */
   public void execute(SqlDropObject drop,
       CalcitePrepare.Context context) {
-    final List<String> path = context.getDefaultSchemaPath();
-    CalciteSchema schema = context.getRootSchema();
-    for (String p : path) {
-      schema = schema.getSubSchema(p, true);
-    }
-    final boolean existed;
+    final Pair<CalciteSchema, String> pair = schema(context, false, drop.name);
+    CalciteSchema schema = pair.left;
+    String objectName = pair.right;
+    assert objectName != null;
+
+    boolean existed;
     switch (drop.getKind()) {
     case DROP_TABLE:
     case DROP_MATERIALIZED_VIEW:
-      existed = schema.removeTable(drop.name.getSimple());
-      if (!existed && !drop.ifExists) {
+      Table materializedView = schema != null && drop.getKind() == SqlKind.DROP_MATERIALIZED_VIEW
+          ? schema.plus().getTable(objectName) : null;
+
+      existed = schema != null && schema.removeTable(objectName);
+      if (existed) {
+        if (materializedView != null) {
+          if (materializedView instanceof Wrapper) {
+            ((Wrapper) materializedView).maybeUnwrap(MaterializationKey.class)
+                .ifPresent(materializationKey -> {
+                  MaterializationService.instance()
+                      .removeMaterialization(materializationKey);
+                });
+          }
+        }
+      } else if (!drop.ifExists) {
         throw SqlUtil.newContextException(drop.name.getParserPosition(),
-            RESOURCE.tableNotFound(drop.name.getSimple()));
+            RESOURCE.tableNotFound(objectName));
       }
       break;
     case DROP_VIEW:
       // Not quite right: removes any other functions with the same name
-      existed = schema.removeFunction(drop.name.getSimple());
+      existed = schema != null && schema.removeFunction(objectName);
```

Comment:
`schema != null` could precondition.

Reply:
Refactored. Note that for every object type we throw an exception with a different error message.